### PR TITLE
Allow Parsers to take Currency object as the forceCurrency parameter

### LINKED ISSUE #386: Parser doesn't accept currency object
Are you willing to accept a pull request to make parser objects accept currency objects in addition to strings?

e.g. this currently throws an error

    $currencies = new ISOCurrencies();
    $moneyParser = new DecimalMoneyParser($currencies);
    $c = new Currency('USD');
    $moneyParser->parse('1.00', $c);

It really should accept the currency object and not force me to write `$c->getCode()` in some places and just `$c` in others.

### PR DESCRIPTION
**Pull request by jarobe42**
Closes #386 

This adds support for $forceCurrency to either be a string or a Currency object.

If you'd rather wait until the next major version and just use the Currency object, let me know and I'll make changes.

### CODE REVIEW

File: src/MoneyParser.php [hunk 1]
```diff
@@ -12,8 +12,8 @@
     /**
      * Parses a string into a Money object (including currency).
      *
-     * @param string      $money
-     * @param string|null $forceCurrency
+     * @param string     $money
+     * @param mixed|null $forceCurrency
```
Comment (sagikazarmark) [inline src/MoneyParser.php]: I would like to see Currency|string|null instead here. Better to be explicit about return types.

File: src/Parser/DecimalMoneyParser.php [hunk 2]
```diff
@@ -46,7 +46,15 @@ public function parse($money, $forceCurrency = null)
             );
         }
 
-        $currency = new Currency($forceCurrency);
+        /*
+         * This conversion is only required whilst currency can be either a string or a
+         * Currency object.
+         */
+        $currency = $forceCurrency;
+        if (!$currency instanceof Currency) {
```
Comment (sagikazarmark) [inline src/Parser/DecimalMoneyParser.php]: I believe we should deprecate passing a string here and remove that in 4.0

So I propose triggering a deprecation notice here:

``` php
@trigger_error('Passing a currency as string is deprecated since 3.1 and will be removed in 4.0. Please pass a '.Currency::class.' instance.', E_USER_DEPRECATED);
```

File: src/Parser/IntlMoneyParser.php [hunk 3]
```diff
@@ -45,23 +45,30 @@ public function parse($money, $forceCurrency = null)
...
-        $decimal = (string) $decimal;
+        if (null !== $forceCurrency) {
+            $currency = $forceCurrency;
+        }
+
+        /*
+         * This conversion is only required whilst currency can be either a string or a
+         * Currency object.
+         */
+        if (!$currency instanceof Currency) {
```
Comment (sagikazarmark) [inline src/Parser/IntlMoneyParser.php]: Same deprecation could go here.

File: src/Parser/IntlMoneyParser.php [hunk 4]
```diff
@@ -45,23 +45,30 @@ public function parse($money, $forceCurrency = null)
...
+            $currency = new Currency($currency);
+        }
+
         $subunit = $this->currencies->subunitFor($currency);
+
+        $decimal = (string) $decimal;
```
Comment (sagikazarmark) [inline src/Parser/IntlMoneyParser.php]: This doesn't seem to be related to the PR. Can you please explain?

File: src/MoneyParser.php (hunk 1 above)
Comment (jarobe42) [reply src/MoneyParser.php]: Good catch. I've updated this.

File: src/Parser/DecimalMoneyParser.php (hunk 2 above)
Comment (jarobe42) [reply src/Parser/DecimalMoneyParser.php]: Good idea. I'll add that in.

File: src/Parser/IntlMoneyParser.php (hunk 4 above)
Comment (jarobe42) [reply src/Parser/IntlMoneyParser.php]: Nothing changed here - I swapped the order of these two following lines, as it made more sense to me that $decimal would be used immediately after being cast;
`$decimal = (string) $decimal;` and
`$subunit = $this->currencies->subunitFor($currency);`

There's no real benefit to doing this though, so I'll flip it back around so as not to add confusion in the git history.

**Comment by jarobe42** [comment]
Thanks for the feedback, @sagikazarmark. I've made the discussed changes, let me know if there's anything else!

File: src/Parser/DecimalMoneyParser.php (hunk 2 above)
Comment (frederikbosch) [reply src/Parser/DecimalMoneyParser.php]: My instinct tells me that is a bad idea. I cannot recall the last time I used `trigger_error`. I rather use the docblock `@deprecated`, which could give visual feedback to developers if their IDE supports it.

Comment (jarobe42) [reply src/Parser/DecimalMoneyParser.php]: `trigger_error` is something I haven't used much lately either, but unless I'm mistaken, there'd be no way to indicate that the use of `parse` is only deprecated when $currency is parsed as a string. I'd be concerned that showing a deprecation notice for all calls of `parse`, including those that pass the correct `Currency` parameter, would be more confusing.

Comment (sagikazarmark) [reply src/Parser/DecimalMoneyParser.php]: This is widely used in the PHP ecosystem. Triggered errors will immediately be bypassed, so no BC break.

@jarobe42 you can and should trigger errors within the if block, only for those calls which received a string, not a currency.

**Comment by sagikazarmark** [comment]
Thanks @jarobe42 a lot, this will be part of the 3.1 release which will happen in the next few weeks.
